File: Data Structures/Linked Lists/Circular Linked Lists/circular-linked-list.js
/* Circular Linked List
*/

function CircularLinkedList() {
  var head = null

  var Node = function(val) {
    this.val = val
    this.next = null
    this.prev = null
  }

  this.head = function() {
    return head
  }

  this.push = function(val) {
    var ptr = new Node(val)
    var temp = head
    ptr.next = head

    if (head !== null) {
      while(temp.next != head) {
        temp = temp.next
      }
      temp.next = ptr
    } else {
      ptr.next = ptr
    }
    head = ptr
  }

  this.remove = function(val){
    var currentNode = head
    var previousNode
    if(currentNode.val === val){
        head = currentNode.next
        currentNode.next.prev = head
    } else {
      while(currentNode.val !== val) {
          previousNode = currentNode
          currentNode = currentNode.next
      }
      previousNode.next = currentNode.next
    }
  }

  this.display = function() {
    var temp = head
    var list = ''
    if (head !== null) {
      while(1) {
        list += temp.val + " => "
        temp = temp.next
        if (temp == head)
            break
      }
    }
    console.log(list)
  }
}

var list = new CircularLinkedList()
list.push('Google')
list.push('Facebook')
list.push('Stack Overflow')
list.push('Gmail')
list.push('Github')
list.remove('Gmail')
list.display()

File: Data Structures/Linked Lists/Circular Linked Lists/circular-linked-list.go
// Double Linked Lists

package main

import (
	"fmt"
)

type Node struct {
	key  interface{}
	next *Node
	prev *Node
}

type CircularLinkedList struct {
	head *Node
}

func (l *CircularLinkedList) Insert(key interface{}) {
	ptr := &Node{
		next: l.head,
		key:  key,
	}
	temp := l.head
	if l.head != nil {
		for temp.next != l.head {
			temp = temp.next
		}
		temp.next = ptr
	} else {
		ptr.next = ptr
	}

	l.head = ptr
}

func (l *CircularLinkedList) Show() {
	CircularLinkedList := l.head
	for CircularLinkedList != nil {
		fmt.Printf("%+v -> ", CircularLinkedList.key)
		CircularLinkedList = CircularLinkedList.next
		if CircularLinkedList == l.head {
			break
		}
	}
	fmt.Println()
}

func main() {
	l := CircularLinkedList{}
	l.Insert(1)
	l.Insert(2)
	l.Insert(3)
	l.Insert(4)
	l.Insert(5)
	l.Insert(6)
	l.Show()
}
